src/skottie-player.ts

```diff
@@ -141,54 +440,368 @@ export class SkottiePlayer extends LitElement {
+        const canvasElement = document.createElement('canvas');
+        canvasElement.setAttribute('id', 'my-canvas');
+        canvasElement.setAttribute('width', this.width.toString());
+        canvasElement.setAttribute('height', this.height.toString());
+
+        const parent = document.getElementById('skottie-player');
+        if (parent) {
+          parent.append(canvasElement);
+          this.canvasSurface = SkottieKit.MakeCanvasSurface('my-canvas');
+        }
+  
+        const surface = (this.canvasSurface as any);
+  
+        const firstFrame = Date.now();
+        const clearColor = SkottieKit.WHITE;
+        const bounds = {fLeft: 0, fTop: 0, fRight: this.width, fBottom: this.height};
+      
+        function drawFrame(canvas: any) {
+            // seek takes a float from 0.0 to 1.0
+            const seek = ((Date.now() - firstFrame) / duration) % 1.0;
+            animation.seek(seek);
+      
+            canvas.clear(clearColor);
+            animation.render(canvas, bounds);
+            surface.requestAnimationFrame(drawFrame);
+        }
+        surface.requestAnimationFrame(drawFrame);
+      });
+    } else {
+      const canvasElement = document.getElementById('my-canvas');
+      if (canvasElement) {
+        canvasElement.setAttribute('style', 'display:none');
+      }
     }
   }
 
   /**
-   * Handles json file upload. Updates content and player.
+   * Handles json file upload. Updates content, editor, and player based on 
+   * file content.
    */
   handleFileUpload(): void {
-    const fileInput = this.shadowRoot?.getElementById('input');
-    if (fileInput) {
-      const files = (fileInput as HTMLInputElement).files;
+    if (this.shadowRoot) {
+      const fileInput = this.shadowRoot.getElementById('input');
+      if (fileInput) {
+        const files = (fileInput as HTMLInputElement).files;
+        if (files) {
+          const file = files[0];
+          if(file) {
+            const reader = new FileReader();
+  
+            this.title = file.name;
+            this.fileUrl = window.URL.createObjectURL(file);
+  
+            reader.readAsText(file);
+            reader.onload = () => {
+              this.setContent(JSON.parse(reader.result as string));
+              this.assets = {};
+              this.imageColors = {};
+              this.fileString = JSON.stringify(this.content);
+              this._loadJSONEditor();
+              this.updatePlayer();
+              this.fileUploaded = true;
+            };
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Handles asset upload. Updates assets, color, and player based on
+   * file content.
+   */
+  handleAssetUpload(e: Event): void {
+    const assetInput = (e.composedPath()[0] as HTMLInputElement);
+    const mapKey = assetInput.getAttribute('map-key');
+    const assetType = assetInput.getAttribute('asset-type')
+    if (assetInput && mapKey && assetType) {
+      const files = assetInput.files;
       if (files) {
```

Comment:
nit: another way you could write this would be to use the typescript optional chaining, you could remove the if statements and make the code overall nicer. just a suggestion
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining

Reply:
Same in the handleFileUpload function